kevm

```diff
@@ -95,30 +102,51 @@ run_interpreter() {
 }
 
 run_test() {
-    local test_file="$1"     ; shift
-    case "$test_file" in
-        *proofs/*     ) run_proof "$test_file" --debug "$@"
-                        ;;
-        *interactive/*) local expected_file="$1" ; shift
-                        local output_file="$(mktemp "$test_file.out.XXXXXX")"
-                        trap "rm -rf $output_file" INT TERM EXIT
-                        run_krun ocaml "$test_file" --interpret "$@" > "$output_file" \
-                            || pretty_diff "$expected_file" "$output_file"
-                        ;;
-        *             ) run_interpreter "$test_file"
-                        ;;
+    local backend=$1        ; shift
+    local test_file="$1"    ; shift
+    case "$backend" in
+        haskell)
+            # To test the haskell backend, compare it to the ocaml backend.
+            local ocaml_output_file="$(mktemp "$test_file.ocaml.out.XXXXXX")"
+            local haskell_output_file="$(mktemp "$test_file.haskell.out.XXXXXX")"
```

Comment:
No need to compare to OCaml backend, instead put a file at `tests/templates/haskell-success.json` and compare to that.

Reply:
I am under the impression that successful tests may produce different outputs. For example, when I test `/tests/ethereum-tests/VMTests/vmBitwiseLogicOperation/gt2.json` the diff is 
```
@@ -1,6 +1,6 @@
 <generatedTop>
   <k>
-    #finalizeTx ( false ) ~> check "gt2" : { "post" : { "0x0f572e5295c57f15886f9b263e2f6d2d6c7b5ec6" : { "balance" : "0xd3c21bcecceda1000000" , "code" : "0x60007fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff11600055" , "nonce" : "0x00" , "storage" : { "0x00" : "0x01" , .JSONList } , .JSONList } , .JSONList } , .JSONList } ~> check "gt2" : { "out" : "0x" , .JSONList } ~> check "gt2" : { "logs" : "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347" , .JSONList } ~> check "gt2" : { "gas" : "0x013874" , .JSONList } ~> clear ~> run { .JSONList } ~> success .EthereumSimulation
+    #finalizeTx ( false ) ~> check "gt2" : { "post" : { "0x0f572e5295c57f15886f9b263e2f6d2d6c7b5ec6" : { "balance" : "0xd3c21bcecceda1000000" , "code" : "0x60007fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff11600055" , "nonce" : "0x00" , "storage" : { "0x00" : "0x01" , .JSONList } , .JSONList } , .JSONList } , .JSONList } ~> check "gt2" : { "out" : "0x" , .JSONList } ~> check "gt2" : { "logs" : "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347" , .JSONList } ~> check "gt2" : { "gas" : "0x013874" , .JSONList } ~> clear ~> run { .JSONList } ~> success ~> .EthereumSimulation
   </k>
   <exit-code>
     1
@@ -14,6 +14,9 @@
   <analysis>
     .Map
   </analysis>
+  <generatedCounter>
+    0
+  </generatedCounter>
 <ethereum>
   <evm>
     <output>
@@ -209,4 +212,3 @@
     </messages>
   </network>
 </ethereum>
-</generatedTop>
```
but when I test `tests/ethereum-tests/VMTests/vmBitwiseLogicOperation/or2.json`
```
@@ -1,6 +1,6 @@
 <generatedTop>
   <k>
-    #finalizeTx ( false ) ~> check "or2" : { "post" : { "0x0f572e5295c57f15886f9b263e2f6d2d6c7b5ec6" : { "balance" : "0xd3c21bcecceda1000000" , "code" : "0x6001600317600055" , "nonce" : "0x00" , "storage" : { "0x00" : "0x03" , .JSONList } , .JSONList } , .JSONList } , .JSONList } ~> check "or2" : { "out" : "0x" , .JSONList } ~> check "or2" : { "logs" : "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347" , .JSONList } ~> check "or2" : { "gas" : "0x013874" , .JSONList } ~> clear ~> run { .JSONList } ~> success .EthereumSimulation
+    #finalizeTx ( false ) ~> check "or2" : { "post" : { "0x0f572e5295c57f15886f9b263e2f6d2d6c7b5ec6" : { "balance" : "0xd3c21bcecceda1000000" , "code" : "0x6001600317600055" , "nonce" : "0x00" , "storage" : { "0x00" : "0x03" , .JSONList } , .JSONList } , .JSONList } , .JSONList } ~> check "or2" : { "out" : "0x" , .JSONList } ~> check "or2" : { "logs" : "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347" , .JSONList } ~> check "or2" : { "gas" : "0x013874" , .JSONList } ~> clear ~> run { .JSONList } ~> success ~> .EthereumSimulation
   </k>
   <exit-code>
     1
@@ -14,6 +14,9 @@
   <analysis>
     .Map
   </analysis>
+  <generatedCounter>
+    0
+  </generatedCounter>
 <ethereum>
   <evm>
     <output>
@@ -209,4 +212,3 @@
     </messages>
   </network>
 </ethereum>
-</generatedTop>
```

In both cases, the only differences are 
- there's a `~>` between `success` and `.EthereumSimulation` at the end of the `<k>` cell
- the extra generated counter cell and the missing `</generatedTop>`
If those differences weren't there, then this would be a successful test. But the contents of the `<k>` cell are different between the two tests. So I don't see how I can have one expected output file for all the tests.

Reply:
The extra `~>` not being added sounds like a bug in the Haskell backend, so it's OK for the tests to fail for that reason.

The extra cell is also a bug in the Haskell backend is my guess, or in the kompile pipeline, so also OK for tests to fail for that reason.

This PR doesn't need to make the tests pass, just make it easier to run the tests. But the intended behavior (for all the backends) is that if a test passes, the entire configuration is emptied out and the exit-code cell is set to `0`. So the correct way to test is to compare against `tests/templates/*-success.json`.

Reply:
Ok, thanks.